Detect duplicate rules for the same 'redirect from' value

sanketio opened the pull request:
<!--
### Requirements

Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

This plugin was allowed to add duplicate redirection rules with same `redirect from` value. This PR will detect the original rule and will redirect to the edit screen of that rule and delete the duplicate rule.

### Benefits

Users now edit the existing redirection rule instead of adding multiple rules with the same `redirect from` value. And also they do not have to remember whether they have added this rule or not, this PR will take care of that and if existing entry found, it'll redirect to the original rule and delete duplicate one.

### Verification Process

Tested manually on multiple websites.

### Applicable Issues

fixes #165

sanketio committed:
Detect duplicate rules for the same 'redirect from' value

jeffpaul committed:
adding trailing semicolon

adamsilverstein commented on inc/classes/class-srm-post-type.php:
```diff
@@ -448,9 +462,50 @@ public function action_save_post( $post_id ) {
+		}
+
+		// Get post if exist with same redirection rule.
+		$existing_post_id = $wpdb->get_var( $wpdb->prepare( "SELECT post_id FROM $wpdb->postmeta WHERE meta_key=%s AND meta_value=%s", '_redirect_rule_from', $redirect_from ) );
```
can we use a WP_Query here instead of a direct $wp_db query? we can limit to fields=['id']

adamsilverstein commented on inc/classes/class-srm-post-type.php:
```diff
@@ -448,9 +462,50 @@ public function action_save_post( $post_id ) {
+	 * @return string
+	 */
+	public function add_notice_query_var( $location, $post_id ) {
+
```
return early if post id not passed or invalid

adamsilverstein commented on inc/classes/class-srm-post-type.php:
```diff
@@ -252,6 +252,20 @@ public function action_redirect_chain_alert() {
+				<?php
+
+				// Delete duplicate post.
+				wp_delete_post( $_GET['duplicate-redirect-rule'], true );
```
Before deleting the post on this passed id, we should add a nonce check, capability check (can the current user delete this?) and sanitize the $_GET variable for security. Otherwise this could be abused.

adamsilverstein reviewed (changes requested):
Nice start on this @sanketio thanks for the PR! I left some small change requests, let me know if you have any questions.

sanketio committed:
Fix feedbacks

- Added nonce and capability check before deleting duplicate redirect rule
- Sanitize variable
- Use WP_Query instead of custom query to fetch the redirect rule with same name

sanketio commented:
@adamsilverstein : Added feedbacks, please check!

adamsilverstein commented on inc/classes/class-srm-post-type.php:
```diff
@@ -448,9 +492,78 @@ public function action_save_post( $post_id ) {
+		}
+
+		$location = add_query_arg(
+			array(
```
I noticed you used `array()` here and `[]` above. Please use one format - whichever best matches what you see elsewhere in the codebase.

adamsilverstein commented:
@sanketio Code looks good now (left one small comment), going to give this a test.

adamsilverstein commented:
@sanketio This worked pretty well in my testing. I have some slight language adjustments I will add as suggestions on the PR. Also, I found one bug, when I hit reload on the page where I am editing the duplicate rule, I get an additional error:

![image](https://user-images.githubusercontent.com/2676022/67112245-fdad7280-f193-11e9-836a-7818d8ac53ee.png)

adamsilverstein commented on inc/classes/class-srm-post-type.php:
```diff
@@ -252,6 +252,50 @@ public function action_redirect_chain_alert() {
+				} elseif ( ! current_user_can( 'delete_post', $duplicate_redirect_rule ) ) {
+
+					$should_delete = false;
+					$error         = __( 'You don\'t have enough permission to delete duplicate redirect rule, please contact your administrator!', 'safe-redirect-manager' );
```
```suggestion
					$error         = __( 'You don\'t have permission to delete the duplicate redirect rule, please contact your administrator!', 'safe-redirect-manager' );
```

adamsilverstein commented on inc/classes/class-srm-post-type.php:
```diff
@@ -252,6 +252,50 @@ public function action_redirect_chain_alert() {
+				?>
+				<div class="notice notice-error is-dismissible">
+					<p>
+						<?php esc_html_e( 'Safe Redirect Manager Error: Duplicate redirection rule found. You can modify below rule instead of adding new one.', 'safe-redirect-manager' ); ?>
```
```suggestion
						<?php esc_html_e( 'Safe Redirect Manager Error: Cannot create rule - duplicate redirection rule found. You can modify the existing rule below.', 'safe-redirect-manager' ); ?>
```

adamsilverstein commented on inc/classes/class-srm-post-type.php:
```diff
@@ -252,6 +252,50 @@ public function action_redirect_chain_alert() {
+				if ( $should_delete ) {
+
+					// Delete duplicate post.
+					wp_delete_post( $duplicate_redirect_rule, true );
```
@sanketio Deleting a post where the ID is passed in a callback seems like a security risk vector I'd love to avoid introducing if possible. 

I'm wondering if this approach could be shifted to acting at a different point, maybe before the rule is even saved? The idea would be:

* On the "Add Redirect" screen, the "Publish" and "Save Draft" buttons would start disabled. 
* As the user (paused/debounced/cached) types into the Redirect From field, send an REST Ajax request to and endpoint that checks for existing matching rules.
* If there is no match, the rule can be published/saved - the buttons become enabled.
* If there is a match, the buttons remain disabled & a warning message appears with a link to "Edit existing rule" (returned by the endpoint if needed).

I realize this adds some code overhead, however if feels like this would result in both a better user experience and a less risky endpoint (no more arbitrary post deleting!) - what do you think?

One more thing: do we want to try to account for wildcards at all? what happens if I have an existing rule for `/foo/*` and I try to add a rule for `foo/bar/` - does that work correctly?

adamsilverstein commented:
@sanketio I love where you are headed in this PR, however the deleting post part got me concerned and I wonder if we can refactor a bit to prevent saving entirely if the rule already exists? See https://github.com/10up/safe-redirect-manager/pull/171#discussion_r336584857

jeffpaul commented:
@sanketio any chance you can respond to the feedback on this PR?

sanketio commented:
@jeffpaul : Sorry :(, I haven't had the time to check this, but as soon as I get some time, I'll definitely try to take over and modify the way as Adam has suggested.

Commit:
Merge branch 'develop' into feature/detect-duplicate-rules

jeffpaul committed:
Merge branch 'develop' into feature/detect-duplicate-rules

lukaspawlik committed:
Merge branch 'develop' into feature/detect-duplicate-rules

dhanendran committed:
text change

dhanendran committed:
using ajax to validate the URL

lukaspawlik commented on inc/classes/class-srm-post-type.php:
```diff
@@ -48,6 +48,8 @@ public function setup() {
 		add_action( 'admin_print_styles-post-new.php', array( $this, 'action_print_logo_css' ), 10, 1 );
 		add_filter( 'post_type_link', array( $this, 'filter_post_type_link' ), 10, 2 );
 		add_filter( 'default_hidden_columns', array( $this, 'filter_hidden_columns' ), 10, 1 );
+		add_action( 'admin_enqueue_scripts', array( $this, 'load_resources' ) );
```
@dhanendran As we're not expecting to get any callback parameters in `load_resources` let's change this as follows

```suggestion
		add_action( 'admin_enqueue_scripts', array( $this, 'load_resources' ), 10, 0 );
```

lukaspawlik commented on inc/classes/class-srm-post-type.php:
```diff
@@ -48,6 +48,8 @@ public function setup() {
 		add_filter( 'post_type_link', array( $this, 'filter_post_type_link' ), 10, 2 );
 		add_filter( 'default_hidden_columns', array( $this, 'filter_hidden_columns' ), 10, 1 );
+		add_action( 'admin_enqueue_scripts', array( $this, 'load_resources' ) );
+		add_action( 'wp_ajax_srm_validate_from_url', array( $this, 'srm_validate_from_url' ) );
```
@dhanendran As we're not expecting to get any callback parameters in `load_resources` let's change this as follows

```suggestion
		add_action( 'wp_ajax_srm_validate_from_url', array( $this, 'srm_validate_from_url' ), 10, 0 );
```

lukaspawlik commented on inc/classes/class-srm-post-type.php:
```diff
@@ -638,4 +641,54 @@ public static function factory() {
+	 */
+	public function load_resources() {
+		if ( 'redirect_rule' === get_post_type() ) {
+			wp_enqueue_script( 'redirectjs', plugin_dir_url( 'safe-redirect-manager/safe-redirect-manager.php' ) . 'assets/js/redirect.js', [ 'jquery' ], SRM_VERSION );
```
```suggestion
			wp_enqueue_script( 'redirectjs', plugin_dir_url( 'safe-redirect-manager/safe-redirect-manager.php' ) . 'assets/js/redirect.js', array( 'jquery' ), SRM_VERSION );
```

lukaspawlik commented on inc/classes/class-srm-post-type.php:
```diff
@@ -638,4 +641,54 @@ public static function factory() {
+		}
+
+		$from = filter_input( INPUT_GET, 'from', FILTER_SANITIZE_STRING );
+		$existing_post_ids = new WP_Query(
```
@dhanendran as we just need to check whether particular post exists, we don't need to populate its meta & term caches so let's ass `update_post_meta_cache` and `update_post_term_cache` to false.

lukaspawlik commented on assets/js/redirect.js:
```diff
@@ -0,0 +1,27 @@
+					$('#message').html( '' ).hide();
+					publishBtn.prop('disabled', false);
+				} else if ( '0' === data ) {
+					$('#message').html( '<p>There are some issues validating the URL. Please try again.</p>' ).show();
```
@dhanendran Let's use localized strings exposed via [wp_localize_script](https://developer.wordpress.org/reference/functions/wp_localize_script/) instead of hardcoding these here and without any translations.

lukaspawlik reviewed (changes requested):
@dhanendran Thanks for your changes! Overall looks good, however I've left you some suggestions and comments.

dhanendran committed:
localize msg and set no. of params for hook

dhanendran commented:
@lukaspawlik I have addressed all your comments. Please review again.

lukaspawlik committed:
Merge branch 'develop' into feature/detect-duplicate-rules

lukaspawlik reviewed (dismissed):
@dhanendran this looks good to me, over to you for next steps!

Sidsector9 committed:
normalize from-to by prepending forward slash if not exist

Sidsector9 reviewed (approved):
Found a minor bug which is fixed in [cca0c1c](https://github.com/10up/safe-redirect-manager/pull/171/commits/cca0c1ca6ab90a572c5e28e555a95ef34550882d), looks good otherwise 🚀
